doc/pins/PINS_SONiC_Design_for_SaiGenericExt.md

```diff
@@ -0,0 +1,429 @@
+
+### P4Orch <a name="P4Orch"></a>
+The P4 Orchagent reads P4RT table entries from APPL_DB, parses, validates and resolves them and then creates necessary SAI objects in ASIC_DB. There are several enhancements to this P4Orch to support the PINS SAI Extension path.
+
+#### Table Definition database <a name="Table-Definition-database"></a>
+This is a database created at the initialization time, extension table entries programming during run-time refers to this data-base. Before processing any extension table entries, P4Orch must have created this database with necessary states like cross table reference, table precedence order for processing APPL_DB entries, mapping each match field and action parameters with associated SAI attribute data types. In PINS architecture, this state is created by processing table definitions supplied by P4RT-APP. P4Orch must have received the **P4RT_TABLE_DEFINITION_SET** entry before P4Orch can process any extension table-entries from APPL_DB. Until this database and necessary state is created, P4Orch should not process any extension table-entries from APPL_DB.
+
+```text
+struct ActionParamInfo
+{
+    std::string     name;
+    std::string     fieldtype;
+    std::string     datatype;
+    std::map<std::string, std::string>  table_reference_map;
+};
+
+struct TableActionInfo
+{
+    std::string     name;
+    std::map<std::string, ActionParamInfo> params;
+};
+
+struct TableMatchInfo
+{
+    std::string        name;
+    std::string        fieldtype;
+    std::string        datatype;
+    std::map<std::string, std::string>  table_reference_map;
+};
+
+struct TableInfo
+{
+    std::string                            name;
+    std::map<std::string, TableMatchInfo>  match_fields;
+    std::map<std::string, TableActionInfo> action_fields;
+    std::vector<std::string>               action_referenced_tables;
+};
+
+typedef std::map<std::string, TableInfo>   TableInfoMap;
+```
+
+#### Demultiplexing PINS SAI Extension entries <a name="Demultiplexing-PINS-SAI-Extension-entries"></a>
+For PINS SAI tables, P4RT-APP publishes related entries in a table with prefix P4RT_TABLE:FIXED_<TableName> (with an exception for ACL table definition). With added support for PINS SAI Extension tables, related entries will be published in a table with prefix **P4RT_TABLE:EXT_<TableName>**.
+
+Table prefix for PINS SAI Path tables is **P4RT_TABLE:FIXED_<TableName>**
+Table prefix for PINS SAI Extension tables would be **P4RT_TABLE:EXT_<TableName>**
+
+Current P4Orch implementation of PINS SAI path maintains mapping from FIXED P4RT table name to the manager responsible for processing respective table entries to relevant SAI Objects. With added support of the PINS SAI Extension tables, control flow for FIXED P4RT tables remains as before, where-as EXT P4RT table entries will be directed to the newly introduced SAI Extension manager.
+
+#### SAI Extension Manager <a name="SAI-Extension-Manager"></a>

```

Comment:
the SAI extension manager looks very much like the SAI-DB itself, except it got pulled up another level higher.

Reply:
Hi Lihua, thank you for your comment. I have some response below. Please see, and we can add more details here as well.